File: services/api/internal/document/domain/document.go
package domain

import (
	"encoding/json"
	"errors"
	"time"

	"github.com/google/uuid"
)

// DocumentType represents the kind of clinical document.
type DocumentType string

const (
	TypeExamRequest  DocumentType = "exam_request"
	TypePrescription DocumentType = "prescription"
	TypeLetter       DocumentType = "letter"
	TypeOther        DocumentType = "other"
)

// DocumentStatus represents the lifecycle state of a clinical document.
type DocumentStatus string

const (
	StatusDraft     DocumentStatus = "draft"
	StatusFinalized DocumentStatus = "finalized"
	StatusPublished DocumentStatus = "published"
)

// Sentinel errors.
var (
	ErrNotFound         = errors.New("document: not found")
	ErrNotDraft         = errors.New("document: operation requires draft status")
	ErrNotFinalized     = errors.New("document: operation requires finalized status")
	ErrAlreadyFinalized = errors.New("document: document is already finalized")
)

// validDocumentTypes holds the set of accepted document types.
var validDocumentTypes = map[DocumentType]bool{
	TypeExamRequest:  true,
	TypePrescription: true,
	TypeLetter:       true,
	TypeOther:        true,
}

// ClinicalDocument represents a clinical document (exam request, prescription, letter, etc.).
type ClinicalDocument struct {
	ID                uuid.UUID
	TenantID          uuid.UUID
	PatientID         uuid.UUID
	ProfessionalID    uuid.UUID
	AppointmentID     *uuid.UUID
	DocumentType      DocumentType
	Title             string
	Status            DocumentStatus
	ContentJSON       json.RawMessage
	VersionNumber     int
	PreviousVersionID *uuid.UUID
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

// Validate checks business rules for a ClinicalDocument.
func (d *ClinicalDocument) Validate() error {
	if d.Title == "" {
		return errors.New("document: title is required")
	}
	if !validDocumentTypes[d.DocumentType] {
		return errors.New("document: invalid document_type")
	}
	return nil
}

// DocumentPublication represents a publication record linking a published document to a patient.
type DocumentPublication struct {
	ID                 uuid.UUID
	TenantID           uuid.UUID
	ClinicalDocumentID uuid.UUID
	PatientID          uuid.UUID
	PublishedByUserID  uuid.UUID
	PublishedAt        time.Time
}

File: packages/api-client/src/types/document.ts
import type { UUID } from "./common"

export type DocumentType = "exam_request" | "prescription" | "letter" | "other"
export type DocumentStatus = "draft" | "finalized" | "published"

export interface ClinicalDocument {
  id: UUID
  patient_id: UUID
  professional_id: UUID
  appointment_id?: string
  document_type: DocumentType
  title: string
  status: DocumentStatus
  content_json: unknown
  version_number: number
  previous_version_id?: string
  created_at: string
  updated_at: string
}

export interface CreateDocumentRequest {
  professional_id: string
  appointment_id?: string
  document_type: DocumentType
  title: string
  content_json: unknown
}

export interface UpdateDocumentRequest {
  title: string
  content_json: unknown
  appointment_id?: string
}
